barbican/plugin/crypto/hsm_partition_crypto.py

```diff
@@ -0,0 +1,232 @@
+
+    This plugin extends the base PKCS11 plugin to support per-project HSM 
+    partitions. Each project is mapped to its own HSM partition with isolated
+    keys and credentials.
+    """
+    
+    def __init__(self, conf=None, ffi=None, pkcs11=None):
+        """Initialize plugin using standard P11CryptoPlugin init."""
+        # Always use our module-level CONF if no config is provided
+        if conf is None:
+            conf = CONF
+ 
+        # Make sure p11_crypto_plugin group is accessible
+        if not hasattr(conf, 'p11_crypto_plugin'):
+            # Register p11 options if not already registered
+            p11_crypto.register_opts(conf)
+
+        # Store partition-specific config
+        self.hsm_partition_conf = conf.hsm_partition_crypto_plugin
+
+        # Initialize basic attributes that parent needs
+        self.library_path = None
+        self.login = None
+        self.rw_session = self.hsm_partition_conf.rw_session
+        self.slot_id = None
+        self.token_labels = None
+        self.token_serial_number = None
+        self.seed_file = self.hsm_partition_conf.seed_file
+        self.seed_length = self.hsm_partition_conf.seed_length
+
+        # Encryption related configs from parent
+        self.encryption_mechanism = self.hsm_partition_conf.encryption_mechanism
+        self.encryption_gen_iv = self.hsm_partition_conf.aes_gcm_generate_iv
+        self.cka_sensitive = self.hsm_partition_conf.always_set_cka_sensitive
+        self.mkek_key_type = 'CKK_AES'  # TODO: Make this also configurable
+        self.mkek_length = self.hsm_partition_conf.mkek_length
+        self.mkek_label = self.hsm_partition_conf.mkek_label
+        self.hmac_key_type = self.hsm_partition_conf.hmac_key_type
+        self.hmac_label = self.hsm_partition_conf.hmac_label
+        self.hmac_mechanism = self.hsm_partition_conf.hmac_mechanism
+        self.key_wrap_mechanism = self.hsm_partition_conf.key_wrap_mechanism
+        self.key_wrap_gen_iv = self.hsm_partition_conf.key_wrap_generate_iv
+        self.os_locking_ok = self.hsm_partition_conf.os_locking_ok
+        self.pkek_length = self.hsm_partition_conf.pkek_length
+        self.pkek_cache_ttl = self.hsm_partition_conf.pkek_cache_ttl
+        self.pkek_cache_limit = self.hsm_partition_conf.pkek_cache_limit
+
+        # Initialize repository interfaces
+        self.hsm_partition_repo = repositories.get_hsm_partition_repository()
+        self.project_hsm_repo = repositories.get_project_hsm_repository()
+
+        # Initialize as None - will be set when first operation occurs
+        # TODO: This implies just one operation at a time
+        self.current_project_id = None
+        self.current_partition = None
+
+        # Create PKCS11 instance
+        # self.pkcs11 = pkcs11 or self._create_pkcs11(ffi)
+    
+        # self._configure_object_cache()
+        # super(HSMPartitionCryptoPlugin, self).__init__(conf, ffi=ffi, pkcs11=None)
+
```

Comment:
to properly register the extension manager needs a friendly name. that he gets from the function `get_plugin_name`
```suggestion
def get_plugin_name(self):
    return self.plugin_name
```
the plugin name needs to be initialized before:
`self.plugin_name = conf.hsm_partition_crypto_plugin.plugin_name`

Reply:
I've added my changes.